application/client/src/app/ui/views/toolbar/chart/zoomer/cursor/state.ts

```diff
@@ -0,0 +1,280 @@
+    }
+
+    public onWheel(event: WheelEvent) {
+        if (this._ng_width === -1 || this._width === -1) {
+            return;
+        }
+        let width: number = this._ng_width;
+        let left: number = 0;
+        // Detect direction
+        if (Math.abs(event.deltaY) > Math.abs(event.deltaX)) {
+            // Vertical scroll: zooming
+            if (event.deltaY < 0) {
+                // Zoom in
+                if (width + event.deltaY < MIN_SIZE_WIDTH) {
+                    width = MIN_SIZE_WIDTH;
+                } else {
+                    width += event.deltaY;
+                }
+            } else if (event.deltaY > 0) {
+                // Zoom out
+                if (width + event.deltaY > this._width) {
+                    width = this._width;
+                } else {
+                    width += event.deltaY;
+                }
+            }
+            left = this._ng_left - Math.round(event.deltaY / 2);
+            if (left < 0) {
+                left = 0;
+            }
+            if (left + width > this._width) {
+                left = this._width - width;
+            }
+        } else {
+            left = this._ng_left + event.deltaX;
+            if (left < 0) {
+                left = 0;
+            }
+            if (left + this._ng_width > this._width) {
+                left = this._width - this._ng_width;
+            }
+        }
+        this._ng_width = width;
+        this._ng_left = left;
+        this._emitChanges();
+        this._parent.detectChanges();
+        this.preventDefault(event);
+    }
+
+    public onClick(event: MouseEvent) {
+        if (this._ng_width === -1 || this._width === -1) {
+            return;
+        }
+        let x: number = event.offsetX;
+        if ((event.target as HTMLElement).className === 'cursor') {
+            x += this._ng_left;
+        }
+        let left: number = x - Math.round(this._ng_width / 2);
+        if (left < 0) {
+            left = 0;
+        }
+        if (left + this._ng_width > this._width) {
+            left = this._width - this._ng_width;
+        }
+        this._ng_left = left;
+        this._emitChanges();
+        this._parent.detectChanges();
+    }
+
+    public onMove(event: MouseEvent) {
+        if (this._ng_width >= this._width) {
+            return;
+        }
+        this._mouse.x = event.x;
+        this._mouse.kind = EChangeKind.move;
+    }
+
+    public onLeft(event: MouseEvent) {
+        this._mouse.x = event.x;
+        this._mouse.kind = EChangeKind.left;
+        event.preventDefault();
+        event.stopImmediatePropagation();
+    }
+
+    public onRight(event: MouseEvent) {
+        this._mouse.x = event.x;
+        this._mouse.kind = EChangeKind.right;
+        event.preventDefault();
+        event.stopImmediatePropagation();
+    }
+
+    public preventDefault(event: MouseEvent) {
+        stop(event);
+        return false;
+    }
```

Comment:
you can use method `stop` from `@ui\env\dom` to stop event. Actually you don't need this method as a part of state.